Make paths also match subdirectories

robwhitaker opened the pull request:
This makes matchers that are generated for strings and paths treat the
path as a prefix instead of as the entire path, allowing a path to also
match its subdirectories and contained files.

This allows patterns like `"src"` to also match by default:
- `src/folder_x`
- `src/folder_x/folder_y`
- etc.

robwhitaker committed:
Make paths also match subdirectories

This makes matchers that are generated for strings and paths treat the
path as a prefix instead of as the entire path, allowing a path to also
match its subdirectories and contained files.

This allows patterns like `"src"` to also match by default:
- `src/folder_x`
- `src/folder_x/folder_y`
- etc.

zimbatm commented:
I would rather add a new matcher than change the default semantic.

In most cases, I want to precisely control all the files that get added. Otherwise, it's too easy for temporary files (generated by the editor or other tooling) to cause unnecessary rebuilds.

zimbatm commented:
Also thanks for your work.

I want to expand a bit on the "Known limitation" section of the README which I believe motivated your work;

The built-in filter asks us to make that decision before entering that folder; if the filter returns false, the filter will stop recursing in that folder. Which is a good thing to avoid traversing parts of the filesystem we know are of no interest (like the .git folder). But sometimes it would be nice to be able to say "maybe", and then only add the folder to the output if any of the files are matching below. That's the limitation, and it's fundamental to how the filter works.

robwhitaker commented:
This PR is addressing the comment, "This probably needs more work, I don't think that it works on sub-folders," more than the limitation noted in the README, though now I think I misunderstood the comment. It doesn't work on sub-folders like `src/some-folder` *because* of the limitation; is that right? It just never descends into `src` in the first place.

That said, I do think it's important to be able to match a folder and its children. It feels a little unintuitive to me that including `"src/"` doesn't include `"src/Main.hs"`, for example, but I understand the reasoning for being conservative here. So maybe the path forward is to:

1. Make an `exactMatch` matcher and clearly document that that is the default behavior if given a raw string or path, and
2. Make this PR into a new matcher (`inDirectory`?)

robwhitaker commented:
Regarding the limitation, I also have another PR in the works addressing that, which I'll open later today. I don't think it's quite ready to merge yet, but we can at least use it as a basis of discussion.

zimbatm commented:
The other issue I had in mind is the reverse of this PR; if the user currently adds "src/sub/folder", it won't automatically match "src" and "src/sub". Currently, the user has to explicitly list all three of them. I think that one is not controversial.

For this PR, like you said there are two options;
* extend the string syntax; this adds a small DSL that has to be documented precisely and that the user has to learn. It could also be extended to match "*.js"-type globs potentially. This wouldn't work on the path syntax.
* add another matcher. This is more verbose but also easier to extend and document. The namespace is much larger in case we want to replace one behavior with another; just add another function.

robwhitaker commented:
> The other issue I had in mind is the reverse of this PR; if the user currently adds "src/sub/folder", it won't automatically match "src" and "src/sub". 

This is actually something I've addressed already in the other PR I haven't opened yet, so I'll hold off on discussing it until that's up.

---

As for this PR, I think the way to go is to leave the default behavior as-is (i.e. `src/` does not automatically match `src/Main.hs`) and turn this PR into a new matcher function (`inDirectory`) that explicitly includes everything within a directory.

I'll push that change sometime this evening.

robwhitaker committed:
Turn _isPathPrefix into new matcher and restore previous default path behavior

zimbatm commented on default.nix:
```diff
@@ -79,4 +84,61 @@ rec {
+    # compare.
+    if builtins.length prefixSegments > builtins.length pathSegments
+    then false
+    else comparison.isPrefix;
```
Nix is not super fast so I try to think about ways to minimize the number of operations that need to happen. Either by memoizing operations or by using faster builtins.

For example here, for each match, both paths will be split (generating more allocations), then each component of the path will be compared.

If we could assume that the paths are "clean" upfront, which is true in most of the cases, then this could be reduced to a string prefix comparison.

robwhitaker replied:
Could you elaborate on what you mean by "clean"?

zimbatm replied:
Sure, a "clean" path is one that is absolute and doesn't contain unnecessary slashes or dots.

If it's possible to assume that, then it's possible to check that

/home/user/src/project/ is a prefix of /home/user/src/project/src/main.js by comparing strings.

The only input that needs to be cleaned is the one passed by the user.

robwhitaker replied:
If I'm not mistaken, Nix paths are clean by default. So, we could either:

a) have the user pass in a path type (enforced via `assert`), or
b) have the matcher automatically turn strings into paths (`root + ("/" + path)`), which would require the matcher to also be passed `root` as a parameter.

I'm more in favor of option (b) here. Much nicer on the user end to write `(inDirectory "src")` than `(inDirectory (root + "/src"))`. Plus, matchers being passed `root` opens them up to being smart about more things.

robwhitaker replied:
Went with option (b), but can change to (a) or something else if you prefer.

zimbatm replied:
Yeah, I was thinking the same yesterday, matchers might want to be initialized with the root.

robwhitaker committed:
Update inDirectory to compare string prefixes instead of path segments

This avoids the extra work of splitting the paths into segments and
comparing them piece by piece.

zimbatm commented on default.nix:
```diff
@@ -31,8 +31,8 @@ rec {
           path__ = "${rootStr}/${f}";
         in
         if builtins.isFunction f then f
```
Same thinking as previously; if the `root` is passed here, the matcher can do pre-processing steps only once rather than on each path.

zimbatm commented on default.nix:
```diff
@@ -43,18 +43,35 @@ rec {
+  inDirectory = directory:
+    root: path: type:
+      let
+        # Convert `directory` to a path to clean user input
+        dirAsPath =
+          if builtins.isString directory then root + ("/" + directory)
+          else if builtins.isPath directory then directory
+          else
+            throw "inDirectory: Unsupported type ${builtins.typeOf directory}, expected string or path";
```
Same thinking as before; this would be checked on each path matches. Unless I missed something, it should be possible to push this to the core filter and do it once, and then provide a clean path guarantee to all the filters.

robwhitaker replied:
Not sure if the conversion could be pushed up to the core filter since the path string is an argument to `inDirectory`. Though we could probably leverage partial application to the same effect. Would something like this work?

In `inDirectory`:

```nix
inDirectory = directory: root:
  let
    dirAsPath =
      if builtins.isString directory then root + ("/" + directory)
      else if builtins.isPath directory then directory
      else
        throw "inDirectory: Unsupported type ${builtins.typeOf directory}, expected string or path";
  in
  path: type:
    let
      directory_ = toString dirAsPath;
      ...
```

In `toMatcher`:

```nix
if builtins.isFunction f then f root
```

zimbatm replied:
Good point.

zimbatm replied:
I pushed the proposed changes as I had them in mind. Let me know what you think!

robwhitaker replied:
Looks mostly good. I don't have time to test much until later, but I caught one error which I commented on.

Good idea moving the path cleaning to a helper function.

zimbatm reviewed (commented):
Ok, this should be the last round of review. The good news is that the other PR will also become trivial to implement after that.

zimbatm committed:
proposed change

robwhitaker commented on default.nix:
```diff
@@ -17,52 +17,45 @@ rec {
+      directory_ = _toCleanPath root directory;
+    in
+    path: type:
+    directory_ == path
```
This will always fail since `directory_` is a string and `path` is a path. Need a `toString` on `path`.

zimbatm replied:
I thought the same as you until I started writing this library. It turns out that the path argument is already a string!

```
nix-repl> builtins.path { path = ./.; filter = path: _: builtins.trace "path=${path} type=${builtins.typeOf path}" false; }  
trace: path=/home/zimbatm/go/src/github.com/numtide/nix-filter/tests/default.nix type=string
trace: path=/home/zimbatm/go/src/github.com/numtide/nix-filter/tests/fixture1 type=string
```

robwhitaker replied:
Huh... good to know!

zimbatm commented:
Thanks a lot for your help, our both brains combined managed to find an elegant solution. I will merge this and deal with the consequences if there are any.
